src/GeniusYield/Transaction.hs

```diff
@@ -488,4 +492,46 @@ makeTransactionBodyAutoBalanceWrapper collaterals ss eh pp ps utxos body changeA
         {- Technically, this doesn't compare with the _final_ tx size, because of signers that will be
         added later. But signing witnesses are only a few bytes, so it's unlikely to be an issue -}
         Left (BuildTxSizeTooBig maxTxSize txSize)
-    first BuildTxCborSimplificationError $ simplifyGYTxBodyCbor $ txBodyFromApi txBody
+
+    collapsedBody <- first BuildTxCollapseExtraOutError $ collapseExtraOut extraOut txBodyContent outsOld
+
+    first BuildTxCborSimplificationError $ simplifyGYTxBodyCbor $ txBodyFromApi collapsedBody
+
+
+{- | Collapses the extra out generated in the last step of tx building into
+    another change output (If one exists)
+-}
+collapseExtraOut
+  :: Api.TxOut Api.S.CtxTx Api.S.BabbageEra
+  -- ^ The extra output generated by makeTransactionBodyAutoBalance
+  -> Api.TxBodyContent Api.S.BuildTx Api.S.BabbageEra
+  -- ^ The body content generted by makeTransactionBodyAutoBalance
+  -> [GYTxOut v]
+  -- ^ The outputs that we don't want to modify
+  -> Either Api.S.TxBodyError (Api.S.TxBody Api.S.BabbageEra)
+  -- ^ The updated body with the collapsed outputs
+collapseExtraOut apiOut@(Api.TxOut outAddr outVal _ _) bodyContent@Api.TxBodyContent {txOuts} outsOld
+ | Api.txOutValueToLovelace outVal == 0 = Api.S.createAndValidateTransactionBody bodyContent
+ | otherwise =
+    case delete apiOut cOuts of
+        [] -> Api.S.createAndValidateTransactionBody bodyContent
+        -- sOut == selected Out == A cOut != eOut that will merge with eOut
+        (sOut@(Api.TxOut sOutAddr sOutVal sOutDat sOutRefScript):_) ->
+            let
+
+            nOutVal = case sOutVal of
+                Api.TxOutAdaOnly oasie l -> Api.TxOutAdaOnly oasie $ Api.txOutValueToLovelace outVal <> l
+                Api.TxOutValue masie val -> Api.TxOutValue masie $ Api.txOutValueToValue outVal <> val
+
+            -- nOut == new Out == The merging of both eOut and sOut
+            nOut = Api.TxOut sOutAddr nOutVal sOutDat sOutRefScript
+            -- nOuts == new Outs == The new list of all old outs mins eOut and sOut plus the nOut
+            nOuts = (txOuts \\ [apiOut, sOut]) ++ [nOut]
+
+            in
+                Api.S.createAndValidateTransactionBody $ bodyContent { Api.txOuts = nOuts }
+  where
+    outsNew = txOuts \\ map txOutToApi outsOld
```

Comment:
This doesn't seem appropriate as it's $O(n^2)$. You are maintaining list of original outputs mentioned in skeleton and filtering against them later. Perhaps what would be appropriate is to maintain list of added change outputs (instead of original skeleton outputs) by our coin selection algorithm (notice the line `(addIns, changeOuts) <- selectInputs`) and then simply merge this additional change output with the last one in it.

Reply:
The use of drop looks fragile to me, because we would be working under the assumption that 
a) the extra outputs are always appended at the end
b) adjustedOuts has the same length as outsOld. 

I took this approach because it was the safest and more robust I could think of without modifying the type of other functions.

If I were to change the type of `balanceTxStep`, I'd just add the changeOuts as another element in the return. Then we can filter those for the change address, just to be sure, and choose an output from there to collapse the extra output into. (Similar to how `makeTransactionBodyAutoBalance` returns the `extraOut` as a separate element, even though it is also included in the body and bodyContent)

I think the amount of outs a tx will ever have is low enough that the time complexity added by the map difference will not be noticeable. But if we are fine changing the type of `balanceTxStep`, the solution described above could be a nice one.

Reply:
> The use of drop looks fragile to me, because we would be working under the assumption that a) the extra outputs are always appended at the end b) adjustedOuts has the same length as outsOld.

These two are invariants which must be respected by our coin selection process and if you think about it, your logic of using `\\` is also relying on these assumptions because `\\` removes for the first match. Secondly, you are matching against old `outsOld` and as we adjust for them later, your match against them may fail.